language/move-ir/types/src/ast.rs

```diff
@@ -555,6 +594,75 @@ pub enum Exp_ {
 /// The type for a `Exp_` and its location
 pub type Exp = Spanned<Exp_>;
 
+//**************************************************************************************************
+// Bytecode
+//**************************************************************************************************
+
+pub type BytecodeBlocks = Vec<(Label, BytecodeBlock)>;
+pub type BytecodeBlock = Vec<Bytecode>;
+
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+pub struct Label(pub String);
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum Bytecode_ {
+    Pop,
+    Ret,
+    BrTrue(Label),
+    BrFalse(Label),
+    Branch(Label),
+    LdU8(u8),
+    LdU64(u64),
+    LdU128(u128),
+    CastU8,
+    CastU64,
+    CastU128,
+    LdByteArray(ByteArray),
+    LdAddr(AccountAddress),
+    LdTrue,
+    LdFalse,
+    CopyLoc(Var),
+    MoveLoc(Var),
+    StLoc(Var),
+    Call(ModuleName, FunctionName, Vec<Type>),
+    Pack(StructName, Vec<Type>),
+    Unpack(StructName, Vec<Type>),
+    ReadRef,
+    WriteRef,
+    FreezeRef,
+    MutBorrowLoc(Var),
+    ImmBorrowLoc(Var),
+    MutBorrowField(StructName, Field),
+    ImmBorrowField(StructName, Field),
+    MutBorrowGlobal(StructName, Vec<Type>),
+    ImmBorrowGlobal(StructName, Vec<Type>),
+    Add,
+    Sub,
+    Mul,
+    Mod,
+    Div,
+    BitOr,
+    BitAnd,
+    Xor,
+    Or,
+    And,
+    Not,
+    Eq,
+    Neq,
+    Lt,
+    Gt,
+    Le,
+    Ge,
+    Abort,
+    GetTxnSenderAddress,
+    Exists(StructName, Vec<Type>),
+    MoveFrom(StructName, Vec<Type>),
+    MoveToSender(StructName, Vec<Type>),
+    Shl,
+    Shr,
+}
+pub type Bytecode = Spanned<Bytecode_>;
```

Comment:
Just to clarify on the usage of Spanned:
- For the incoming assembly-like IR, it is easy to map a bytecode to a source location.
- However for compiler generated code, what do we expect the span to be? Does it map back to a source location in the Move code? I also wouldn't be surprised if in some cases the source compiler will insert a few instructions to make things work and curious what the spans would be.

Reply:
All source location information is now shared #2758